auth: added user authentication and authorization.

FaiyazMujawar opened the pull request:
Fixes #6 

**User authentication**

I've added user authentication and authorization with JWT. The user can send their credentials through `POST` method on `api/auth/login`. Upon successful validation of user credentials, the user is responded back with a JWT access token in JSON fomat.

The JWT token payload contains the userID and the user department.

**Format of response on successful validation:**
```
{
    "token":<token>
}
```

**Adding new user:**

New User can be added by sending the required user data in `x-www-form-urlencoded` format using `POST` method on `api/users/add`. 

**Required fields:**
`zprn`         :  String - 7 characters long, containing all numbers
`firstName` :  String 
`lastName` :  String
`password` :  String - atleast 6 characters long
`department` : String - atleast 3 characters long
`role`          :  String

**Updating user password:**
User can update their password by sending the old password and new password on `api/users/update`, along with the valid JWT access token, using `POST` request. The provided old password is matched with the password in the database. Upon successful match, the new password is hashed and stored in the database

drigio commented on models/User.js:
```diff
@@ -0,0 +1,31 @@
+
+const UserSchema = new mongoose.Schema({
+    zprn: {
+        type: String,
```
Need to change this to number as you are validating only for the digits in `Joi`

FaiyazMujawar replied:
I'll see to this

drigio commented on routes/auth.js:
```diff
@@ -0,0 +1,69 @@
+                            });
+
+                            token != null
+                                ? res.status(200).json({ token: token })
```
We need to also add a `result` parameter that can have to values only - `failure` and `success`. So that when the api call is returned it is easy to understand whether to consider it as success or failure and read the error. We need to add this param in `Error` as well which will default to `result: "failure"`.

FaiyazMujawar replied:
Okay

drigio commented on routes/auth.js:
```diff
@@ -0,0 +1,69 @@
+ */
+function validateCredentials(zprn, password) {
+    const schema = Joi.object({
+        zprn: Joi.string().required().length(7),
```
Add here the digit regex as well to check only for a number in `zprn`.

drigio commented on routes/user.js:
```diff
@@ -0,0 +1,165 @@
+const userService = require("../services/UserService");
+
+router
+    .route("/")
```
We need to give it a proper route maybe? `api/users/` seems like we are `GET`ing a list of `users` rather than adding a new user. We need to have an endpoint like `api/users/add` or something like that.

drigio commented on routes/user.js:
```diff
@@ -0,0 +1,165 @@
+                });
+        }
+    })
+    .patch((req, res) => {
```
Also shouldn't we have like a parameter to update a user in the `url`? Accepting it in body seems different. 
Check out [this](https://zulipchat.com/api/update-user) link for how we can also do it. `api/user/{user_id}` is the url that will be more suitable?

drigio commented on services/UserService.js:
```diff
@@ -0,0 +1,82 @@
+    // UPDATE USER PASSWORD
+    async updateUser(zprn, password) {
+        return new Promise((resolve, reject) => {
+            User.findOneAndUpdate(
```
`updateUser` shouldn't only update the user's password. We should update the whole `user` object.  We should only pass parameters that we want to change

drigio reviewed (changes requested):
I've put comments here. Check out [this](https://zulip.readthedocs.io/en/latest/git/fixing-commits.html#changing-the-last-commit) link to fix the last commit.

FaiyazMujawar committed:
auth: added user authentication and authorization.

Added user authentication and authorization with JWT. The user can send
their credentials through POST method on api/auth/login. Upon successful
validation of user credentials, the user is responded back with a JWT access
token in JSON fomat.The JWT token payload contains the userID and the user 
department.

New User can be added by sending the required user data in x-www-form-urlencoded
format using POST method on api/users/add.

User can update their password by sending the old password and new password on 
api/users/update, along with the valid JWT access token, using POST request. The
provided old password is matched with the password in the database. Upon 
successful match, the new password is hashed and stored in the database

drigio commented:
You don't need to do `git push` after every `git commit` you do. Make all the changes locally and then push it once you are ready for the final one.
